app/controllers/AccountingPeriodController.scala

```diff
@@ -80,15 +80,31 @@ class AccountingPeriodController @Inject() (
             )
 
             for {
-              updatedAnswers <- Future.fromTry(request.userAnswers match {
-                                  case Some(answers) =>
-                                    answers.set(AccountingPeriodPage, formWithAccountingPeriodEnd)
-                                  case None =>
-                                    UserAnswers(request.userId).set(AccountingPeriodPage, formWithAccountingPeriodEnd)
-                                })
+              (updatedAnswers, dynamicMode) <- Future.fromTry(request.userAnswers match {
+                                                 case Some(answers) =>
+                                                   val maybePrevAnswer = answers.get(AccountingPeriodPage)
+
+                                                   val answerChanged = maybePrevAnswer.forall {
+                                                     case prevAnswer if prevAnswer == formWithAccountingPeriodEnd =>
+                                                       false
+                                                     case _ => true
+                                                   }
+
+                                                   val mode = if (answerChanged) NormalMode else CheckMode
+
+                                                   val finalAnswers = if (answerChanged) answers.clean else answers
+
+                                                   finalAnswers
+                                                     .set(AccountingPeriodPage, formWithAccountingPeriodEnd)
+                                                     .map(_ -> mode)
+                                                 case None =>
+                                                   UserAnswers(request.userId)
+                                                     .set(AccountingPeriodPage, formWithAccountingPeriodEnd)
+                                                     .map(_ -> mode)
```

Comment:
Instead of a lot of intermediate variables, you could replace with an expression like ...

```
for {
              (updatedAnswers, mode) <- Future.fromTry(request.userAnswers match {
                                          case Some(answers) =>
                                            answers
                                              .get(AccountingPeriodPage)
                                              .map {
                                                case prevAnswer if prevAnswer == formWithAccountingPeriodEnd =>
                                                  Success((answers, CheckMode))
                                                case _ =>
                                                  answers
                                                    .clean
                                                    .set(AccountingPeriodPage, formWithAccountingPeriodEnd)
                                                    .map(_ -> NormalMode)
                                              }
                                              .getOrElse {
                                                answers
                                                  .set(AccountingPeriodPage, formWithAccountingPeriodEnd)
                                                  .map(_ -> mode)
                                              }
                                          case None =>
                                            UserAnswers(request.userId)
                                              .set(AccountingPeriodPage, formWithAccountingPeriodEnd)
                                              .map(_ -> mode)
                                        })
              _ <- sessionRepository.set(updatedAnswers)
            }
```